indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorStateManager.java

```diff
@@ -0,0 +1,386 @@
+          UNABLE_TO_CONNECT_TO_STREAM,
+          LOST_CONTACT_WITH_STREAM
+      );
+      return !unhealthyStates.contains(this);
+    }
+  }
+
+  public enum StreamErrorTransience
+  {
+    TRANSIENT,
+    POSSIBLY_TRANSIENT,
+    NON_TRANSIENT,
+    NON_STREAM_ERROR
+  }
+
+  private State supervisorState;
+  private final int healthinessThreshold;
+  private final int unhealthinessThreshold;
+  private final int healthinessTaskThreshold;
+  private final int unhealthinessTaskThreshold;
+
+  private boolean atLeastOneSuccessfulRun = false;
+  private boolean currentRunSuccessful = true;
+  private final CircularBuffer<TaskState> completedTaskHistory;
+  private final CircularBuffer<State> supervisorStateHistory; // From previous runs
+  private final ExceptionEventStore eventStore;
+
+  public SeekableStreamSupervisorStateManager(
+      State initialState,
+      SeekableStreamSupervisorConfig supervisorConfig
+  )
+  {
+    this.supervisorState = initialState;
+    this.healthinessThreshold = supervisorConfig.getHealthinessThreshold();
+    this.unhealthinessThreshold = supervisorConfig.getUnhealthinessThreshold();
+    this.healthinessTaskThreshold = supervisorConfig.getTaskHealthinessThreshold();
+    this.unhealthinessTaskThreshold = supervisorConfig.getTaskUnhealthinessThreshold();
+    Preconditions.checkArgument(
+        supervisorConfig.getMaxStoredExceptionEvents() >=
+        Math.max(healthinessThreshold, unhealthinessThreshold),
+        "numExceptionEventsToStore must be greater than or equal to both "
+        + "healthinessThreshold and unhealthinessThreshold"
+    );
+    this.eventStore = new ExceptionEventStore(
+        supervisorConfig.getMaxStoredExceptionEvents(),
+        supervisorConfig.isStoringStackTraces()
+    );
+    this.completedTaskHistory = new CircularBuffer<>(Math.max(healthinessTaskThreshold, unhealthinessTaskThreshold));
+    this.supervisorStateHistory = new CircularBuffer<>(Math.max(healthinessThreshold, unhealthinessThreshold));
+  }
+
+  /**
+   * Certain supervisor states can only be set if the supervisor hasn't had a successful iteration yet.  This function
+   * checks if there's been at least one successful iteration if needed and sets supervisor state to an appropriate
+   * new state.
+   */
+  public Optional<State> setState(State newState)
+  {
+    if (newState.isOnlySetWhenNoSuccessfulRunYet()) {
+      if (!atLeastOneSuccessfulRun) {
+        supervisorState = newState;
+        return Optional.of(newState);
+      } else {
+        return Optional.absent();
+      }
+    } else if (newState.equals(State.SUSPENDED)) {
+      atLeastOneSuccessfulRun = false; // We want the startup states again after being suspended
+    }
+    this.supervisorState = newState;
+    return Optional.of(newState);
+  }
+
+  public void storeThrowableEvent(Throwable t)
+  {
+    if (t instanceof PossiblyTransientStreamException && atLeastOneSuccessfulRun) {
+      t = new TransientStreamException(t);
```

Comment:
How about wrap the underlying exception instead of wrapping `PossiblyTransientStreamException` in a `TransientStreamException`?